### ISSUE TITLE
Fleet Movement Error Results HTTP 500 Error

### ISSUE THREAD

**Issue by asparatu** [XGProyect/XG-Proyect-v3.x.x#448]
**Describe the bug**
There is still an HTTP 500 Error that happens when you have Fleet movements. I believe has only do when transport resource to plant or moon.  This error does not happen all the time, but somehow the mission does not get sometimes meaning instead of the value of 3 being set in the fleet table it is zero. I have debugged the whole fleet movement from start to finish there is no place I can see that updates the fleet_mission field expect when inserted. This makes me believe it does not get a set mission value.

app/controllers/game/fleet4.php Line 561
` // add the fleet mission
$this->_fleet_data['fleet_mission'] = $data['mission'];`

I believe the $data['mission'] is not being set.

**Screenshots**
Database record showing fleet_mission is zero
![fleet_movement_error](https://user-images.githubusercontent.com/3848896/103163830-42c04080-483e-11eb-9919-23316ea9be87.jpg)

Debug Error shows the same thing saying offset is zero
![fleet_movement_offset_error](https://user-images.githubusercontent.com/3848896/103163854-8024ce00-483e-11eb-925d-ca5ed551025d.jpg)

When I change the fleet_mission to 3 the HTTP 500 Error is gone and the game portal works again and the fleet movement is completed and deleted from the server.

I would suggest you add value 0 and just call it default or something stop this errro from happening.

 **Server (please complete the following information):**
 - PHP Version: PHP 7.3
 - MySQL Version:  MySQL 5.7

**Comment by duhow** [comment]
Can you provide all steps to reproduce the issue?
Also what XGP version / commit are you using?

**Comment by asparatu** [comment]
@duhow I am using version 3.2.0 and I created a fleet that 5000 large cargo ships, sent to my moon, and add some resources about 20k, 40k, 1 million. I select the transport. 

This fleet movement issue does not happen all the time, so I am not sure what happens.

**Comment by BeReal86** [comment]
Hi @asparatu, Thank you for your feedback with priority.

I am using php 7.4 version. When I tried on localhost, I did not get any errors. The XGP version I use is in the link below, as you mentioned.

https://github.com/XGProyect/XG-Proyect-v3.x.x/releases/tag/v3.2.0

I will experiment on hosting.

**Comment by asparatu** [comment]
@BeReal86 

I found a way to reproduce the fleet movement error. When you get the page that you select the fleet movement if you do not select anything the option will set as zero. 

For example, if you just add the resources and don't select transportation or deployment option when press continue it saves to the database. when the fleet gets to where suppose to the code does not know what do and crashes and give the HTTP 500 error because there is no fleet movement number defined. 

So, what I can see is there is validation before you save to the database because if there was this error would not happen. This why I said does not happen all the time.

**Comment by asparatu** [comment]
Hello @BeReal86 

Here is the code I use to fix the problem in the control/game/fleet4.php. This will stop from continuing and not put zero in the fleet_mission database column.

```
//add if statement to check mission status is zero redirect to fleet1 page 
        if ($data['mission'] == 0) {
            Functions::redirect('game.php?page=fleet1');
         } else {
            // add the fleet mission
            $this->_fleet_data['fleet_mission'] = $data['mission'];
         }
```

**Comment by duhow** [comment]
@asparatu do you mean to set that code in here?
https://github.com/XGProyect/XG-Proyect-v3.x.x/blob/52a10d352076a50b10c0840316dc4aca4e3a1c75/src/upload/app/controllers/game/fleet4.php#L245-L247

**Comment by asparatu** [comment]
Hello @duhow 

That will not work. What you just did was check to  see if the whole array is null and it will return false everytime because there is data in $data array already from the previous pages.

If where to do that you need to do
if(is_null($data['mission'])) {
Functions::redirect('game.php?page=fleet1')
}

You need only to check if the mission is null and nothing else because you are setting that variable.

**Comment by LucasKovacs** [comment]
fixed on 3.3 - thanks for everybody's contribution
